### PR TITLE
Make vbguest plugin usable in combination with registration plugin

### PR DESCRIPTION
**Pull request by hferentschik**

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -3,11 +3,14 @@ source 'https://rubygems.org'
...
+  gem 'vagrant',      git: 'https://github.com/mitchellh/vagrant.git'
+  gem 'vagrant-spec', git: 'https://github.com/mitchellh/vagrant-spec.git'
   gem 'gem-compare'
   gem 'rake'
+  gem 'vagrant-vbguest'
 end
 
 group :plugins do
   gem 'vagrant-registration', path: '.'
+  gem 'pry-byebug'
```
Comment (pvalena) [inline Gemfile]: We certainly don't want to include dependency on 'pry-byebug'.

File: Gemfile.lock [hunk 2]
```diff
@@ -0,0 +1,158 @@
+GIT
```
Comment (pvalena) [inline Gemfile.lock]: Is the .lock file necessary?

File: lib/vagrant-registration/action/register.rb [hunk 3]
```diff
@@ -42,7 +37,8 @@ def call(env)
 
           @logger.debug('Registration is skipped due to the configuration') if config.skip
 
-          @app.call(env) if Plugin.vbguest_plugin?
+          # Call next middleware in chain
+          @app.call(env)
```
Comment (pvalena) [inline lib/vagrant-registration/action/register.rb]: This cannot be done, as it does not work with libvirt provider: http://paste.fedoraproject.org/343422/58251114

However the hook and the detection of VirtualBox provider could work. I will test that tomorrow.

**Comment by pvalena** [comment]
Can you please put the tests into separate PR?

File: Gemfile (hunk 1 above)
Comment (hferentschik) [inline Gemfile]: It's a def dependency. Happy to remove it though, since it is hard do step into vagrant-spec execution.

File: Gemfile.lock (hunk 2 above)
Comment (hferentschik) [inline Gemfile.lock]: Yes, absolutely. The lock file is essential to the whole process of using Bundler. Only when you check in the lock file you guarantee repeatable builds/setups. There is a argument for not checkin it in for library gems, but imo it is still better to have it checked in and then do dependencies upgrades at dedicated points in time.

File: lib/vagrant-registration/action/register.rb (hunk 3 above)
Comment (hferentschik) [inline lib/vagrant-registration/action/register.rb]: > This cannot be done, as it does not work with libvirt provider:

Sure, you need to call the next middleware in chain. I guess the problem is in this case in _plugin.rb_. I'll change back to adding the registration middleware after folder sync for not virtual box.

> However using the different hook for VirtualBox provider could probably work.

I know it does. That's what the tests are for. It should be very easy to write/eable the same tests against libvirt as well.

**Comment by hferentschik** [comment]
> Can you please put the tests into separate PR?

How come? Tests and functionality change go hand in hand imo. I guess in this case where  a new testing framework is added, I can separate them.

File: Gemfile.lock (hunk 2 above)
Comment (hferentschik) [inline Gemfile.lock]: I kept the commit separate so that it can easily be dropped in case we don't want to check it in.

File: lib/vagrant-registration/action/register.rb (hunk 3 above)
Comment (hferentschik) [inline lib/vagrant-registration/action/register.rb]: > I guess the problem is in this case in plugin.rb. 

Or maybe not!? Not quite sure yet why the comm layer should not be up.

**Comment by hferentschik** [comment]
@pvalena would you mind to try this updated pull request again against libvirt?

File: Gemfile.lock (hunk 2 above)
Comment (pvalena) [inline Gemfile.lock]: Ok, thank you.

**Comment by hferentschik** [comment]
@pvalena Got a Linux box as well now. I will test against libvirt (changing the harness to be runnable against libvirt as well). On Linux using VirtualBox all looks good as well.

**Comment by pvalena** [comment]
> How come? Tests and functionality change go hand in hand imo. I guess in this case where a new testing framework is added, I can separate them.

Sorry, I wasn't very clear on this. Yes, I meant it like a general addition of testing framework, including the tests of the vbguest-issue, that should already be solved (it's working on Linux). It is an OSX-specific issue.

> @pvalena Got a Linux box as well now. I will test against libvirt (changing the harness to be runnable against libvirt as well). On Linux using VirtualBox all looks good as well.

Good to hear, as right now I am unable to test libvirt and I won't have much time next week (I can run the tests though).

**Comment by hferentschik** [comment]
Ok, I pushed an update. Now with tests for VirtualBox and Libvirt. To run the tests you need to set the following variables:

* VAGRANT_SPEC_BOX_VIRTUAL_BOX - (file) URL to the VirtualBox box for testing
* VAGRANT_SPEC_BOX_LIBVIRT - (file) URL to the Libvirt box for testing
* SUBSCRIPTION_USER - Valid Red Hat subscription user
* SUBSCRIPTION_PASSWORD - Valid Red Hat subscription password

You need to set/provide at least one of the BOX URLs. The other two variables are mandatory.

Then you can run the tests via (after a `bundle install` of course):

    vagrant-spec test --components provider/virtualbox/cdk
    vagrant-spec test --components provider/libvirt/cdk

BTW, I really like how the vagrant-spec tests are shaping up. They are slow, but at least I can test the current code without building a gem first and they are also true acceptance tests.

**Comment by LalatenduMohanty** [comment]
I have built vagrant-registration plugin with this PR and installed on OS X along with vagrant-vbguest plugin. Got below error during `vagrant up`. However it tried to register the guest. Not sure anything wrong with the network as I was on a public network.
```
Lalatendus-MacBook-Pro:cdk20-beta5 lmohanty$ vagrant up
Bringing machine 'default' up with 'virtualbox' provider...
==> default: Importing base box 'cdk20beta5'...
==> default: Matching MAC address for NAT networking...
==> default: Setting the name of the VM: cdk20-beta5_default_1458880995361_39898
==> default: Clearing any previously set network interfaces...
==> default: When using virtualbox, a non-NAT network interface is required.
==> default: Adding a private network using DHCP
==> default: Preparing network interfaces based on configuration...
    default: Adapter 1: nat
    default: Adapter 2: hostonly
==> default: Forwarding ports...
    default: 22 => 2222 (adapter 1)
==> default: Booting VM...
==> default: Waiting for machine to boot. This may take a few minutes...
    default: SSH address: 127.0.0.1:2222
    default: SSH username: vagrant
    default: SSH auth method: private key
    default: Warning: Connection timeout. Retrying...
    default: Warning: Remote connection disconnect. Retrying...
    default: 
    default: Vagrant insecure key detected. Vagrant will automatically replace
    default: this with a newly generated keypair for better security.
    default: 
    default: Inserting generated public key within guest...
    default: Removing insecure key from the guest if it's present...
    default: Key inserted! Disconnecting and reconnecting using new SSH key...
==> default: Machine booted and ready!
==> default: Registering box with vagrant-registration...
    default: Would you like to register the system now (default: yes)? [y|n]y
    default: username: lmohanty@redhat.com
    default: password: 
No installation found.
Loaded plugins: product-id, search-disabled-repos, subscription-manager
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.


 One of the configured repositories failed (Red Hat Enterprise Linux 7 Server - Extended Update Support (RPMs)),
 and yum doesn't have enough cached data to continue. At this point the only
 safe thing yum can do is fail. There are a few ways to work "fix" this:

     1. Contact the upstream for the repository and get them to fix the problem.

     2. Reconfigure the baseurl/etc. for the repository, to point to a working
        upstream. This is most often useful if you are using a newer
        distribution release than is supported by the repository (and the
        packages for the previous distribution release still work).

     3. Disable the repository, so yum won't use it by default. Yum will then
        just ignore the repository until you permanently enable it again or use
        --enablerepo for temporary usage:

            yum-config-manager --disable rhel-7-server-eus-rpms

     4. Configure the failing repository to be skipped, if it is unavailable.
        Note that yum will try to contact the repo. when it runs most commands,
        so will have to try and fail each time (and thus. yum will be be much
        slower). If it is a very temporary problem though, this is often a nice
        compromise:

            yum-config-manager --save --setopt=rhel-7-server-eus-rpms.skip_if_unavailable=true

failure: repodata/repomd.xml from rhel-7-server-eus-rpms: [Errno 256] No more mirrors to try.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
==> default: Checking for guest additions in VM...
    default: No guest additions were detected on the base box for this VM! Guest
    default: additions are required for forwarded ports, shared folders, host only
    default: networking, and more. If SSH fails on this machine, please install
    default: the guest additions and repackage the box to continue.
    default: 
    default: This is not an error message; everything may continue to work properly,
    default: in which case you may ignore this message.
The following SSH command responded with a non-zero exit status.
Vagrant assumes that this means the command failed!

yum install -y kernel-devel-`uname -r` gcc make perl bzip2

Stdout from the command:

Loaded plugins: product-id, search-disabled-repos, subscription-manager


Stderr from the command:

https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
Trying other mirror.


 One of the configured repositories failed (Red Hat Enterprise Linux 7 Server - Extended Update Support (RPMs)),
 and yum doesn't have enough cached data to continue. At this point the only
 safe thing yum can do is fail. There are a few ways to work "fix" this:

     1. Contact the upstream for the repository and get them to fix the problem.

     2. Reconfigure the baseurl/etc. for the repository, to point to a working
        upstream. This is most often useful if you are using a newer
        distribution release than is supported by the repository (and the
        packages for the previous distribution release still work).

     3. Disable the repository, so yum won't use it by default. Yum will then
        just ignore the repository until you permanently enable it again or use
        --enablerepo for temporary usage:

            yum-config-manager --disable rhel-7-server-eus-rpms

     4. Configure the failing repository to be skipped, if it is unavailable.
        Note that yum will try to contact the repo. when it runs most commands,
        so will have to try and fail each time (and thus. yum will be be much
        slower). If it is a very temporary problem though, this is often a nice
        compromise:

            yum-config-manager --save --setopt=rhel-7-server-eus-rpms.skip_if_unavailable=true

failure: repodata/repomd.xml from rhel-7-server-eus-rpms: [Errno 256] No more mirrors to try.
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
https://cdn.redhat.com/content/eus/rhel/server/7/7Server/x86_64/os/repodata/repomd.xml: [Errno 14] curl#7 - "Failed to connect to 64:ff9b::5f65:40fb: Network is unreachable"
```
```
$ vagrant plugin list
landrush (0.19.0)
vagrant-hostmanager (1.5.0)
  - Version Constraint: 1.5.0
vagrant-registration (1.2.1)
  - Version Constraint: 1.2.1
vagrant-service-manager (0.0.4)
  - Version Constraint: 0.0.4
vagrant-share (1.1.5, system)
vagrant-triggers (0.5.2)
vagrant-vbguest (0.11.0)
```

File: Gemfile.lock (hunk 2 above)
Comment (LalatenduMohanty) [inline Gemfile.lock]: +1 for committing the .lock file. It helps backtrack the exact version number if you get in to gem version issues.

File: lib/vagrant-registration/action/register.rb (hunk 3 above)
Comment (LalatenduMohanty) [inline lib/vagrant-registration/action/register.rb]: > This cannot be done, as it does not work with libvirt provider: http://paste.fedoraproject.org/343422/58251114

@hferentschik  @pvalena Is this output from the vagrant-spec based tests?

**Comment by hferentschik** [comment]
>  However it tried to register the guest. Not sure anything wrong with the network as I was on a public network.

That look very muck like a network error. Note, the problem this pull request addresses is even solved in your case. If you look at the output, you can see that the registration occurred prior to vbguest trying to install the guest additions. This is what was not working before. For all I can tell, vbguest additions was then not able to reach the repositories needed to get the packages it needs.

File: lib/vagrant-registration/action/register.rb (hunk 3 above)
Comment (hferentschik) [inline lib/vagrant-registration/action/register.rb]: To be clear, there was indeed an issue with the initial change I made, but I addressed this in a following push. What I wanted to achieve was to consistently call the registration on the "down" hook. The existing implementation would call the registration on the "down" hook for everything except libvirt and for libvirt it would call it on the uphook. That does not make sense. You really want to make wait until the machine is reachable and then just call registration and the next middleware in the chain (aka using the downhook).

The old implementation was hooking into FolderSync for libvirt, which for some providers means the machine is already reachable, however, in the case of libvirt this is not the case. Hence, you needed to to the register on the "up" hook. I changed the code in _plugin.rb_, so that libvirt registers an action after it is known that the machine is reachable. Now, the registration itself can always occur on the "down" hook. 

Bottom line, there was initially an issue with the pull request,  but that's addressed.

File: Gemfile.lock (hunk 2 above)
Comment (hferentschik) [inline Gemfile.lock]: It is the only way to ensure consistent setups. As said, there is a point to be made about libraries development, but imo checking in the lock file is always the better approach.

File: README.md [hunk 4]
```diff
@@ -316,6 +361,27 @@ environment variable on `1` before executing the test script:
...
+
+```
+vagrant-spec test --components provider/virtualbox/cdk
+vagrant-spec test --components provider/libvirt/cdk
+```
+
+In order to successfully run the tests, you need to set the following
+environment variables:
+
+* _VAGRANT_SPEC_BOX_VIRTUAL_BOX_ - URL to the VirtualBox box for testing
```
Comment (LalatenduMohanty) [inline README.md]: @hferentschik This means an user has to download the CDK Vgrantbox from customer portal and then host it in some http location. Do we have anyway to specify if the Vagrantbox is already present on the test machine? May be a file path or config.vm.box name if the box is already installed on the  system.

Comment (hferentschik) [inline README.md]: You already can use a file path. Just use (file://<path to the box file). That's how I do it as well and how it is intended. Downloading the box on each test would take so long, especially since the box won't be cached.

As a dev you just download the boxes you are interested to your preferred location and point the variables to it. Then you can run the tests against these boxes as often as you like. I envision this to be used in a CI environment as well. The build setup would copy the boxes to a dedicated location (probably within the checkout) and then just set the appropriate environment variables.

Comment (hferentschik) [inline README.md]: So what's the status here? Can we get this merged?

**Comment by hferentschik** [comment]
I think I might change this pull request to also use Cucumber + Aruba for the acceptance tests, instead of vagrant-spec!? Thoughts?

**Comment by LalatenduMohanty** [comment]
@hferentschik It is up to you as I think you have better understanding of things at this point of time. So if you think  Cucumber + Aruba is a better way to do acceptance tests then lets do it.

**Comment by hferentschik** [comment]
If we merge the Cucumber + Aruba for vagrant-service-manager, I think it makes sense to align. Most importantly though I would like to get the actual functional change in, since I think it is sound.

**Comment by hferentschik** [comment]
Don't merge right now. I will push an update later today switching to Cucumber + Aruba

**Comment by hferentschik** [comment]
Good to go now.

**Comment by LalatenduMohanty** [comment]
Tested it on OSX + Virtualbox. The Patch resolves the conflict and vagrant-vbguest installed the Virtualbox guest additions. Test also passed successfully.

```
Lalatendus-MacBook-Pro:adb-vagrant-registration lmohanty$ bundle exec rake features PROVIDER=virtualbox
[DEPRECATION] `last_comment` is deprecated.  Please use `last_description` instead.
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby -S bundle exec cucumber 
Using the default and html profiles...
............----....----....----

8 scenarios (3 skipped, 5 passed)
32 steps (12 skipped, 20 passed)
9m56.823s
```

**Comment by hferentschik** [comment]
> Tested it on OSX + Virtualbox. The Patch resolves the conflict and vagrant-vbguest installed the Virtualbox guest additions. Test also passed successfully.

Nice.

**Comment by hferentschik** [comment]
@LalatenduMohanty I went ahead and merged this.
